src/web_interface/templates/upload/upload.html

```diff
@@ -30,229 +37,212 @@ <h3 class="mb-3">Upload Firmware</h3>
         <div id="form">
             <form class="form-horizontal" action="" method=post enctype=multipart/form-data>
 
-                <label class="control-label" for="selectFile">File:</label>
-                <div class="form-group">
-                    <div{% if 'file' in error.keys() %} class="has-error"{% endif %}>
-                        <input class="form-control-file" id="selectFile" type="file" name=file required>
-                        {% if "file" in error.keys() %}
-                            <span class="help-block">{{ error["file"] }}</span>
-                        {% endif %}
-                    </div>
-                </div>
+                {# File #}
+                {%- if not is_reanalysis %}
+                    {% call macros.upload_input("file", "File", error) %}
+                        <input class="form-control-file" id="file" type="file" name=file required>
+                    {% endcall %}
+                {%- else %}
+                    <input type="hidden" name="file_name" id="file_name" value="{{ firmware.file_name }}">
+                {%- endif %}
 
                 <script>
-                    var device_class_dropdown = document.getElementById('device_class_dropdown');
-                    var vendor_dropdown = document.getElementById('vendor_dropdown');
+                    var device_names = eval( {{ device_names | safe }} );
+
                     function add_device_class_options(selected_device_class, selected_vendor, data) {
-                        $('#device_name_dropdown').empty();
+                        $('#device_name').empty();
                         if (data.hasOwnProperty(selected_device_class)) {
                             if (data[selected_device_class].hasOwnProperty(selected_vendor)) {
                                 var device_classes = data[selected_device_class][selected_vendor];
                                 device_classes.sort();
                                 for (var key in device_classes) {
                                     if (device_classes.hasOwnProperty(key)) {
-                                        $('#device_name_dropdown').append('<option>' + device_classes[key] + '</option>');
+                                        $('#device_name').append('<option>' + device_classes[key] + '</option>');
                                     }
                                 }
                             }
                         }
-                        $('#device_name_dropdown').append('<option>' + 'new entry' + '</option>');
+                        $('#device_name').append('<option>' + 'new entry' + '</option>');
```

Comment:
At about this point in the file I now decided to not code review this file further.
I'll make a functional review and leave it at that.